### ISSUE TITLE
Deconstructing a particular shape of complex type crashes the compiler

### ISSUE THREAD

**Issue by airbreather** [dotnet/roslyn#50654]
## Version Used
Visual Studio Enterprise 2019, 16.8.4

## Steps to Reproduce
1. Load the following project in Visual Studio.
1. Move the text cursor to immediately before the keyword "`var`".
1. Open the quick actions menu.

Edit: going through the original version where this showed up, I actually wrote a comment indicating that `var` is actually the workaround for the underlying bug, which is that this pattern actually stops the compiler in its tracks.  This specific set of repro steps happens to just be a very tame way to observe the behavior in an easily repeatable fashion; it's probably not a problem with the code fix provider itself.  I have updated the title accordingly.

### Program.cs
```csharp
(int, (int, (int, int), (int, int)))[] vals = null;
foreach (var (a, (b, (c, d), (e, f))) in vals)
{
}
```

### ConsoleApp0.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

</Project>
```

## Expected Behavior
The "Use explicit type instead of 'var'" action can turn this:
```csharp
foreach (var (a, (b, (c, d), (e, f))) in vals)
{
}
```

into this:
```csharp
foreach ((int a, (int b, (int c, int d), (int e, int f))) in vals)
{
}
```

## Actual Behavior
`UseExplicitTypeCodeFixProvider` raises an unexpected exception with the following stack trace:
```
System.ArgumentOutOfRangeException : Specified argument was out of the range of valid values.
Parameter name: index
   at Microsoft.CodeAnalysis.SeparatedSyntaxList`1.get_Item(Int32 index)
   at Microsoft.CodeAnalysis.CSharp.Binder.MakeDeconstructionConversion(TypeSymbol type,SyntaxNode syntax,SyntaxNode rightSyntax,DiagnosticBag diagnostics,ArrayBuilder`1 variables,Conversion& conversion)
   at Microsoft.CodeAnalysis.CSharp.Binder.MakeDeconstructionConversion(TypeSymbol type,SyntaxNode syntax,SyntaxNode rightSyntax,DiagnosticBag diagnostics,ArrayBuilder`1 variables,Conversion& conversion)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindDeconstructionAssignment(CSharpSyntaxNode node,ExpressionSyntax left,BoundExpression boundRHS,ArrayBuilder`1 checkedVariables,Boolean resultIsUsed,DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindDeconstruction(CSharpSyntaxNode deconstruction,ExpressionSyntax left,ExpressionSyntax right,DiagnosticBag diagnostics,DeclarationExpressionSyntax& declaration,ExpressionSyntax& expression,Boolean resultIsUsedOverride,BoundDeconstructValuePlaceholder rightPlaceholder)
   at Microsoft.CodeAnalysis.CSharp.ForEachLoopBinder.BindForEachPartsWorker(DiagnosticBag diagnostics,Binder originalBinder)
   at Microsoft.CodeAnalysis.CSharp.ForEachLoopBinder.BindForEachParts(DiagnosticBag diagnostics,Binder originalBinder)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindForEachParts(DiagnosticBag diagnostics,Binder originalBinder)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindForEachParts(DiagnosticBag diagnostics,Binder originalBinder)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindForEach(CommonForEachStatementSyntax node,DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindStatement(StatementSyntax node,DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.IncrementalBinder.BindStatement(StatementSyntax node,DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.Bind(Binder binder,CSharpSyntaxNode node,DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.MethodBodySemanticModel.Bind(Binder binder,CSharpSyntaxNode node,DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetBoundNodes(CSharpSyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetLowerBoundNode(CSharpSyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetBoundNodes(CSharpSyntaxNode node,CSharpSyntaxNode& bindableNode,BoundNode& lowestBoundNode,BoundNode& highestBoundNode,BoundNode& boundParent)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetSymbolInfoWorker(CSharpSyntaxNode node,SymbolInfoOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetSymbolInfoWorker(CSharpSyntaxNode node,SymbolInfoOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSymbolInfo(ExpressionSyntax expression,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSymbolInfoFromNode(SyntaxNode node,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSymbolInfoCore(SyntaxNode node,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Simplification.SimplificationHelpers.GetOriginalSymbolInfo(SemanticModel semanticModel,SyntaxNode expression)
   at Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers.NameSimplifier.TrySimplify(NameSyntax name,SemanticModel semanticModel,OptionSet optionSet,TypeSyntax& replacementNode,TextSpan& issueSpan,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers.ExpressionSimplifier.TryReduceExplicitName(ExpressionSyntax expression,SemanticModel semanticModel,TypeSyntax& replacementNode,TextSpan& issueSpan,OptionSet optionSet,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers.ExpressionSimplifier.TrySimplify(ExpressionSyntax expression,SemanticModel semanticModel,OptionSet optionSet,ExpressionSyntax& replacementNode,TextSpan& issueSpan,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Simplification.CSharpNameReducer.SimplifyName(SyntaxNode node,SemanticModel semanticModel,OptionSet optionSet,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Simplification.AbstractCSharpReducer.AbstractReductionRewriter.SimplifyNode[TNode](TNode node,SyntaxNode newNode,SyntaxNode parentNode,Func`5 simplifier)
   at Microsoft.CodeAnalysis.CSharp.Simplification.AbstractCSharpReducer.AbstractReductionRewriter.SimplifyExpression[TExpression](TExpression expression,SyntaxNode newNode,Func`5 simplifier)
   at Microsoft.CodeAnalysis.CSharp.Simplification.CSharpNameReducer.Rewriter.VisitIdentifierName(IdentifierNameSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.IdentifierNameSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitAliasQualifiedName(AliasQualifiedNameSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Simplification.CSharpNameReducer.Rewriter.VisitAliasQualifiedName(AliasQualifiedNameSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.AliasQualifiedNameSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitQualifiedName(QualifiedNameSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Simplification.CSharpNameReducer.Rewriter.VisitQualifiedName(QualifiedNameSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.QualifiedNameSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitDeclarationExpression(DeclarationExpressionSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.DeclarationExpressionSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitArgument(ArgumentSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitListElement[TNode](TNode node)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitList[TNode](SeparatedSyntaxList`1 list)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitTupleExpression(TupleExpressionSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.TupleExpressionSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitForEachVariableStatement(ForEachVariableStatementSyntax node)
   at Microsoft.CodeAnalysis.CSharp.Syntax.ForEachVariableStatementSyntax.Accept[TResult](CSharpSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.Visit(SyntaxNode node)
   at Microsoft.CodeAnalysis.CSharp.Simplification.AbstractCSharpReducer.AbstractReductionRewriter.VisitNodeOrToken(SyntaxNodeOrToken nodeOrToken,SemanticModel semanticModel,Boolean simplifyAllDescendants)
   at async Microsoft.CodeAnalysis.Simplification.AbstractSimplificationService`3.<>c__DisplayClass10_1.<ReduceAsync>b__0[TExpressionSyntax,TStatementSyntax,TCrefSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Simplification.AbstractSimplificationService`3.ReduceCoreAsync[TExpressionSyntax,TStatementSyntax,TCrefSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Simplification.AbstractSimplificationService`3.ReduceAsync[TExpressionSyntax,TStatementSyntax,TCrefSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Simplification.Simplifier.ReduceAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.CleanupDocumentAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.PostProcessChangesAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.PostProcessAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetPreviewOperationsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.GetPreviewResultAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.<>c__DisplayClass11_0.<GetPreviewAsync>b__0(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunctionAsync[T](<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

**Comment by airbreather** [comment]
I suppose I should also mention a point that's probably also important, though I suspect that you'll get there without this comment to help: if I actually do the code fix manually, Visual Studio raises a huge ruckus, so it's probably something related to what the result of the code fix would be:
![image](https://user-images.githubusercontent.com/3792815/105246253-f916fd00-5b40-11eb-8cd2-ae6a93f76ab9.png)

**Comment by jcouv** [comment]
Fix is going into 16.10
